src/Rules/Register.hs

```diff
@@ -1,44 +1,101 @@
-module Rules.Register (registerPackage) where
+{-# LANGUAGE TupleSections #-}
+module Rules.Register (registerPackages) where
 
 import Base
 import Context
 import GHC
 import Target
 import Utilities
+import Hadrian.Expression
+import Settings
 
--- TODO: Simplify.
+import Distribution.ParseUtils
+import qualified Distribution.Compat.ReadP as Parse
+import Distribution.Version (Version)
+
+import Hadrian.Haskell.Cabal.Parse as Cabal
+
+parseCabalName :: String -> Maybe (String, Version)
+parseCabalName = readPToMaybe parse
+  where parse = (,) <$> (parsePackageName <* Parse.char '-') <*> parseOptVersion
+
+-- | This rule provides rules for copying packges into the
+-- boot packages db from the installed compiler.
 -- | Build rules for registering packages and initialising package databases
 -- by running the @ghc-pkg@ utility.
-registerPackage :: [(Resource, Int)] -> Context -> Rules ()
-registerPackage rs context@Context {..} = do
-    when (stage == Stage0) $ do
-        -- Packages @ghc-boot@ and @ghc-boot-th@ both match the @ghc-boot*@
-        -- pattern, therefore we need to use priorities to match the right rule.
-        -- TODO: Get rid of this hack.
-        "//" ++ stage0PackageDbDir -/- pkgName package ++ "*.conf" %%>
-            buildConf rs context
+registerPackages :: [(Resource, Int)] -> Context -> Rules ()
+registerPackages rs context@Context {..} = do
+    root <- buildRootRules
+    root -/- inplacePackageDbPath stage %>
+      buildStamp rs context
+
+    root -/- inplacePackageDbPath stage -/- packageDbStamp %> \stamp -> do
```

Comment:
What is going on here? Surely we need to do *something* besides just writing the stamp file, no?

Reply:
This is a shake short coming iirc. You can not *depend* on folders. You can only depend on files. As such, depending on the `stamp` file, will implicitly build the folder, and call the rule above, which does the initialization.

Reply:
Where is this rule `need`ed?